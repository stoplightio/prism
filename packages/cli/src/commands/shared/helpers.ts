import { getHttpOperationsFromResource } from '@stoplight/prism-http';
import * as signale from 'signale';
import { createMultiProcessPrism, CreatePrismOptions, createSingleProcessPrism } from '../../util/createServer';

export const middleware = async (argv: any) => (argv.operations = await getHttpOperationsFromResource(argv.document!));

export const onFail = (yargs: any, msg: string, err: { message: string }) => {
  if (msg) {
    yargs.showHelp();
  } else {
    signale.fatal(err.message);
  }

  process.exit(1);
};

export const documentPositional: [string, object] = [
  'document',
  {
    description: 'Path to a document. Can be both a file or a fetchable resource on the web.',
    type: 'string',
  },
];

export const handler = (opts: any, parsedArgs: any) => {
  const { multiprocess, port, host, cors, operations } = (parsedArgs as unknown) as CreatePrismOptions & {
    multiprocess: boolean;
  };

  return (multiprocess ? createMultiProcessPrism : createSingleProcessPrism)({ cors, port, host, operations, ...opts });
};
