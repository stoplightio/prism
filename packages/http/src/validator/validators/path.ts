import { HttpParamStyles, IHttpPathParam } from '@stoplight/types';
import { IHttpNameValue } from '../../types';
import { IHttpParamDeserializerRegistry } from '../deserializers/types';
import { HttpParamsValidator } from './params';

export class HttpPathValidator extends HttpParamsValidator<IHttpNameValue> {
  constructor(
    registry: IHttpParamDeserializerRegistry<IHttpNameValue>,
    prefix: string,
    style: HttpParamStyles = HttpParamStyles.Simple
  ) {
    super(registry, prefix, style);
  }
  public validate(target: IHttpNameValue, specs: IHttpPathParam[]) {
    return super.validate(target, specs);
  }
}
