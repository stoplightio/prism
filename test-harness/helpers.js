const fs = require('fs');
const { join } = require('path');
const fetch = require('node-fetch');

const getPort = ({ PRISM_PORT }) => PRISM_PORT  || 4010;

async function makeRequest({ path, method, headers = {}, body }) {
  const opts =
    method === 'GET' || method === 'HEAD'
      ? {}
      : { body: headers['Content-Type'] === 'application/json' ? JSON.stringify(body) : body };
  const baseOpts = Object.assign({}, opts, { method, headers });
  const host = `http://localhost:${getPort(process)}`;
  const requestConfig = {
    ...baseOpts,
    path,
    host,
  };

  return fetch(`${host}${path}`, requestConfig)
    .then(async response => {
      const { date, ...headers } = response.headers.raw();

      return {
        request: requestConfig,
        response: {
          status: response.status,
          statusText: response.statusText,
          headers: headers,
          body: await response.json(),
        },
      };
    })
    .catch(err => {
      console.log('err', err);
    });
}

function constructMasterFileName(request) {
  return JSON.stringify(request).replace(/[{},":/]/gim, '_');
}

function readFile(hash) {
  const fileContent = fs.readFileSync(join(__dirname, '/gold-master-files/') + `${hash}.json`).toString();

  return JSON.parse(fileContent);
}

module.exports = {
  constructMasterFileName,
  makeRequest,
  readFile,
  getPort
};
