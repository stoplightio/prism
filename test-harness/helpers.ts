import { Dictionary } from '@stoplight/types/dist';
import * as xmlDiff from 'diff-js-xml';
import * as parser from 'fast-xml-parser';
import * as typeIs from 'type-is';

type Result = { body: string; headers: Dictionary<string> };

export const xmlValidator = {
  test: (contentType: string, content: string) => {
    const doesContentTypeMatch = !!typeIs.is(contentType, [
      'application/xml',
      'application/*+xml',
      'text/xml',
    ]);
    const isContentXML = parser.validate(content) === true;

    return doesContentTypeMatch || isContentXML;
  },
  validate: (expected: Result, output: Result) => {
    return new Promise(res =>
      xmlDiff.diffAsXml(expected.body, output.body, {}, { compareElementValues: false }, result =>
        res(result)
      )
    );
  },
};

export function parseSpecFile(spec: string) {
  const regex = /====(server|test|spec|command|expect|expect-loose)====\r?\n/gi;
  const splitted = spec.split(regex);

  const testIndex = splitted.findIndex(t => t === 'test');
  const specIndex = splitted.findIndex(t => t === 'spec');
  const serverIndex = splitted.findIndex(t => t === 'server');
  const commandIndex = splitted.findIndex(t => t === 'command');
  const expectIndex = splitted.findIndex(t => t === 'expect');
  const expectLooseIndex = splitted.findIndex(t => t === 'expect-loose');

  return {
    test: splitted[1 + testIndex],
    spec: splitted[1 + specIndex],
    server: splitted[1 + serverIndex],
    command: splitted[1 + commandIndex],
    expect: splitted[1 + expectIndex],
    expectLoose: splitted[1 + expectLooseIndex],
  };
}
