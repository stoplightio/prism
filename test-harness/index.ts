import { ChildProcess, spawn, spawnSync } from 'child_process';
import * as fs from 'fs';
import { validate } from 'gavel';
import glob = require('glob');
import { parseResponse } from 'http-string-parser';
import { get } from 'lodash';
import * as path from 'path';
import * as split2 from 'split2';
import * as tmp from 'tmp';
import { parseSpecFile, xmlValidator } from './helpers';

jest.setTimeout(15000);

const WAIT_FOR_LINE = 'Prism is listening';
const WAIT_FOR_LINE_TIMEOUT = 10000;

describe('harness', () => {
  const files = process.env.TESTS
    ? String(process.env.TESTS).split(',')
    : glob.sync('**/*.txt', { cwd: path.join(__dirname, './specs') });

  files.forEach(file => {
    const data = fs.readFileSync(path.join(__dirname, './specs/', file), { encoding: 'utf8' });
    const parsed = parseSpecFile(data);

    let tmpFileHandle: tmp.FileSyncObject;

    beforeAll(() => {
      tmpFileHandle = tmp.fileSync({
        postfix: '.yml',
        dir: undefined,
        name: undefined,
        prefix: undefined,
        tries: 10,
        template: undefined,
        unsafeCleanup: undefined,
      });

      fs.writeFileSync(tmpFileHandle.name, parsed.spec, { encoding: 'utf8' });
    });

    afterAll(() => tmpFileHandle.removeCallback(undefined, undefined, undefined, undefined));
    describe(file, () => {
      let prismHandle: ChildProcess;
      beforeEach(async () => {
        prismHandle = await startPrism(parsed.server, tmpFileHandle.name);
      });

      afterEach(() => {
        return shutdownPrism(prismHandle);
      });

      test(parsed.test, async () => {
        const [command, ...args] = parsed.command.split(/ +/).map(t => t.trim());

        const clientCommandHandle = spawnSync(command, args, {
          shell: true,
          encoding: 'utf8',
          windowsVerbatimArguments: false,
        });
        const output: any = parseResponse(clientCommandHandle.stdout.trim());
        const expected: any = parseResponse((parsed.expect || parsed.expectLoose || parsed.expectKeysOnly).trim());

        const isXml = xmlValidator.test(get(output, ['header', 'content-type'], ''), expected.body);

        if (isXml) {
          const res = await xmlValidator.validate(expected, output);
          expect(res).toStrictEqual([]);
          delete expected.body;
          delete output.body;

          const isValid = validate(expected, output).valid;
          expect(isValid).toBeTruthy();
          return;
        }

        const isValid = validate(expected, output).valid;

        if (!!isValid) {
          expect(isValid).toBeTruthy();
        } else {
          expect(output).toMatchObject(expected);
        }
        if (parsed.expect) {
          expect(output.body).toStrictEqual(expected.body);
        } else if (parsed.expectKeysOnly) {
          const jsonOutput = JSON.parse(output.body);
          const jsonExpected = JSON.parse(expected.body);
          expect(Object.keys(jsonOutput)).toStrictEqual(Object.keys(jsonExpected));
        }
      });
    });
  });
});

function startPrism(server: string, filename: string): Promise<ChildProcess> {
  return new Promise((resolve, reject) => {
    const serverArgs = server.split(/ +/).map(t => t.trim().replace('${document}', filename));
    const prismMockProcessHandle = spawn(path.join(__dirname, '../cli-binaries/prism-cli'), serverArgs);

    const timeout = setTimeout(() => {
      shutdownPrism(prismMockProcessHandle);
      reject(new Error(`Timeout while waiting for "${WAIT_FOR_LINE}" log line`));
    }, WAIT_FOR_LINE_TIMEOUT);

    if (process.env.DEBUG) {
      prismMockProcessHandle.stderr.pipe(process.stderr);
    }

    prismMockProcessHandle.stdout.pipe(split2()).on('data', (line: string) => {
      if (line.includes(WAIT_FOR_LINE)) {
        clearTimeout(timeout);
        resolve(prismMockProcessHandle);
      }
    });
  });
}

function shutdownPrism(processHandle: ChildProcess): Promise<void> {
  processHandle.kill();
  return new Promise(resolve => {
    processHandle.on('exit', resolve);
  });
}
