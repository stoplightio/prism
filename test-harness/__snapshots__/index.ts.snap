// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`harness When I send a request to an operation with Accept header containing application/json
And this operation can produce a response that matches this X format
Then I should get back a response matching that format
And this response should have Content-Type header matching X format
And this response should have 200 status code
 1`] = `
Snapshot Diff:
- First value
+ Second value

- HTTP/1.1 200 OK
- content-type: application/json; charset=utf-8
- content-length: 7
- Connection: keep-alive
- 
+ HTTP/1.1 200 OK
+ content-type: application/json; charset=utf-8
+ content-length: 7
+ Connection: keep-alive
+ 
  "hello"
`;
